File: frontend/src/types/domain.ts
/** Domain contracts mirrored from backend Pydantic schemas. No business logic here. */

export type SensorType = "sentinel-2" | "sentinel-1";
export type DataMode = "development" | "earth_engine";
export type AnalysisStatus = "pending" | "running" | "completed" | "failed";
export type TraceStatus = "pending" | "running" | "completed" | "failed";

export interface GeoJSONGeometry {
  type: "Polygon" | "MultiPolygon" | "Point" | "LineString";
  coordinates: number[][][] | number[][] | number[];
}

export interface AOI {
  geometry: GeoJSONGeometry;
  name?: string | null;
  area_km2?: number | null;
}

export interface ImageryPreferences {
  cloud_cover_max?: number;
  prefer_least_cloud?: boolean;
}

export interface ImageryRequest {
  aoi: AOI;
  start_date: string;
  end_date: string;
  sensor?: SensorType;
  preferences?: ImageryPreferences;
}

export interface ImageryScene {
  scene_id: string;
  acquisition_date: string;
  cloud_cover_percent?: number | null;
  preview_url?: string | null;
}

export interface SpatialMetadata {
  crs?: string;
  bbox: number[];
  resolution_m?: number | null;
}

export interface ImageryResult {
  source: string;
  mode: DataMode;
  sensor: SensorType;
  scenes: ImageryScene[];
  spatial: SpatialMetadata;
  message?: string | null;
}

export interface QueryRequest {
  query: string;
  aoi?: AOI;
  earlier_date?: string;
  later_date?: string;
  sensor?: SensorType;
  preferences?: ImageryPreferences;
  image_id?: string;
  earlier_image_id?: string;
  later_image_id?: string;
  optical_image_id?: string;
  sar_image_id?: string;
}

export type ImageModality = "optical" | "multispectral" | "sar";
export type ImageFormat = "geotiff" | "tiff" | "png" | "jpeg";

export interface ImageInput {
  id: string;
  modality: ImageModality;
  format: ImageFormat;
  filename: string;
  width: number;
  height: number;
  georeferenced: boolean;
  acquisition_datetime?: string | null;
  benchmark_dataset?: boolean;
  co_registered_benchmark?: boolean;
  benchmark_pair_id?: string | null;
  bounds?: number[] | null;
  crs?: string | null;
}

export interface UploadImageResponse {
  image: ImageInput;
}

export interface SingleImageVQAResult {
  task: "single_image_vqa";
  answer: string;
  model_name: string;
  model_version: string;
  provider: "development" | "geochat_service";
  provenance: string;
  confidence?: number | null;
  confidence_available: boolean;
  input_image_id: string;
  requested_modality: string;
  inference_metadata?: Record<string, unknown>;
}

export interface SingleImageCaptionResult {
  task: "single_image_caption";
  description: string;
  model_name: string;
  model_version: string;
  provider: "development" | "geochat_service";
  provenance: string;
  confidence?: number | null;
  confidence_available: boolean;
  input_image_id: string;
  requested_modality: string;
  inference_metadata?: Record<string, unknown>;
}

export interface CrossModalOpticalSARResult {
  task: "cross_modal_optical_sar";
  answer: string;
  question: string;
  optical_analysis: {
    modality: string;
    summary: string;
    analyzer: string;
    provider: string;
    confidence_available: boolean;
  };
  sar_analysis: {
    modality: string;
    summary: string;
    analyzer: string;
    provider: string;
    confidence_available: boolean;
  };
  fused_analysis: {
    summary: string;
    fusion_policy: string;
    fused_region_count: number;
    complementary_notes: string[];
  };
  co_registration_status: string;
  co_registration_provenance: string;
  optical_image_id: string;
  sar_image_id: string;
  provider: string;
  provenance: string;
  confidence?: number | null;
  confidence_available: boolean;
}

export interface BiTemporalChangeResult {
  task: "bi_temporal_change_vqa";
  change_summary: string;
  question: string;
  changed_region_count: number;
  change_map_available: boolean;
  detector: string;
  provider: "development" | "uploaded_cva";
  provenance: string;
  confidence?: number | null;
  confidence_available: boolean;
  earlier_image_id: string;
  later_image_id: string;
  earlier_acquisition: string;
  later_acquisition: string;
  earlier_date: string;
  later_date: string;
  inference_metadata?: Record<string, unknown>;
}

export interface Metric {
  name: string;
  value: number | string;
  unit?: string | null;
  source: string;
}

export interface EvidenceRegion {
  id: string;
  geometry: GeoJSONGeometry;
  type: string;
  confidence: number;
  metrics: Metric[];
  source: string;
  metadata: Record<string, unknown>;
}

export interface PlanQueryMetadata {
  planner?: string;
  intent?: string;
  required_tools?: string[];
  requested_modalities?: string[];
  planner_version?: string;
  fallback_used?: boolean;
  status?: string;
  duration_ms?: number;
}

export interface TraceStep {
  id: string;
  tool_name: string;
  status: TraceStatus;
  started_at?: string | null;
  completed_at?: string | null;
  duration_ms?: number | null;
  summary?: string | null;
  error?: string | null;
  metadata?: PlanQueryMetadata | Record<string, unknown> | null;
}

export interface AnalysisResult {
  status: AnalysisStatus;
  session_id: string;
  answer: string;
  confidence: number;
  confidence_available?: boolean;
  metrics: Metric[];
  evidence: EvidenceRegion[];
  trace: TraceStep[];
  mode: DataMode;
  vqa?: SingleImageVQAResult | null;
  caption?: SingleImageCaptionResult | null;
  bi_temporal_change?: BiTemporalChangeResult | null;
  cross_modal?: CrossModalOpticalSARResult | null;
}

export interface ApiResponse<T> {
  success: boolean;
  data: T;
}

export interface SubmitQueryData {
  session_id: string;
  result: AnalysisResult;
}

export interface ErrorResponse {
  success: false;
  error: { code: string; message: string; field?: string | null };
}

File: backend/app/schemas/domain.py
from __future__ import annotations

from datetime import date, datetime
from enum import Enum
from typing import Any, Literal

from pydantic import BaseModel, Field, field_validator, model_validator


class SensorType(str, Enum):
    SENTINEL_2 = "sentinel-2"
    SENTINEL_1 = "sentinel-1"


class DataMode(str, Enum):
    DEVELOPMENT = "development"
    EARTH_ENGINE = "earth_engine"


class AnalysisStatus(str, Enum):
    PENDING = "pending"
    RUNNING = "running"
    COMPLETED = "completed"
    FAILED = "failed"


class TraceStatus(str, Enum):
    PENDING = "pending"
    RUNNING = "running"
    COMPLETED = "completed"
    FAILED = "failed"


class GeoJSONGeometry(BaseModel):
    type: Literal["Polygon", "MultiPolygon", "Point", "LineString"]
    coordinates: list[Any]

    @field_validator("coordinates")
    @classmethod
    def coordinates_not_empty(cls, v: list[Any]) -> list[Any]:
        if not v:
            raise ValueError("geometry coordinates must not be empty")
        return v


class AOI(BaseModel):
    """Area of interest as GeoJSON geometry with optional metadata."""

    geometry: GeoJSONGeometry
    name: str | None = None
    area_km2: float | None = Field(default=None, ge=0)

    @model_validator(mode="after")
    def validate_polygon(self) -> AOI:
        if self.geometry.type not in ("Polygon", "MultiPolygon"):
            raise ValueError("AOI geometry must be Polygon or MultiPolygon")
        return self


class ImageryPreferences(BaseModel):
    cloud_cover_max: float = Field(default=30.0, ge=0, le=100)
    prefer_least_cloud: bool = True


class ImageryRequest(BaseModel):
    aoi: AOI
    start_date: date
    end_date: date
    sensor: SensorType = SensorType.SENTINEL_2
    preferences: ImageryPreferences = Field(default_factory=ImageryPreferences)

    @model_validator(mode="after")
    def validate_dates(self) -> ImageryRequest:
        if self.end_date < self.start_date:
            raise ValueError("end_date must be on or after start_date")
        return self


class SpatialMetadata(BaseModel):
    crs: str = "EPSG:4326"
    bbox: list[float] = Field(description="[min_lon, min_lat, max_lon, max_lat]")
    resolution_m: float | None = None


class ImageryScene(BaseModel):
    scene_id: str
    acquisition_date: date
    cloud_cover_percent: float | None = None
    preview_url: str | None = None
    platform_id: str | None = Field(
        default=None,
        description="Provider-native image identifier for downstream analysis (e.g. EE asset path)",
    )
    metadata: dict[str, Any] = Field(
        default_factory=dict,
        description="Provider-specific scene metadata (bands, tile, path/row, etc.)",
    )


class ImageryResult(BaseModel):
    source: str
    mode: DataMode
    sensor: SensorType
    scenes: list[ImageryScene]
    spatial: SpatialMetadata
    collection_id: str | None = Field(
        default=None,
        description="Source image collection identifier (e.g. COPERNICUS/S2_SR_HARMONIZED)",
    )
    provider_metadata: dict[str, Any] = Field(
        default_factory=dict,
        description="Provider-level metadata (selection policy, project, etc.)",
    )
    message: str | None = Field(
        default=None,
        description="Human-readable note, e.g. development adapter disclaimer",
    )


class QueryRequest(BaseModel):
    query: str = Field(min_length=3, max_length=2000)
    aoi: AOI
    earlier_date: date
    later_date: date
    sensor: SensorType = SensorType.SENTINEL_2
    preferences: ImageryPreferences = Field(default_factory=ImageryPreferences)

    @model_validator(mode="after")
    def validate_query_dates(self) -> QueryRequest:
        if self.later_date <= self.earlier_date:
            raise ValueError("later_date must be after earlier_date")
        return self


class Metric(BaseModel):
    name: str
    value: float | int | str
    unit: str | None = None
    source: str


class EvidenceRegion(BaseModel):
    id: str
    geometry: GeoJSONGeometry
    type: str = "change"
    confidence: float = Field(ge=0, le=1)
    metrics: list[Metric] = Field(default_factory=list)
    source: str
    metadata: dict[str, Any] = Field(default_factory=dict)


class TraceStep(BaseModel):
    id: str
    tool_name: str
    status: TraceStatus
    started_at: datetime | None = None
    completed_at: datetime | None = None
    duration_ms: int | None = None
    summary: str | None = None
    error: str | None = None
    metadata: dict[str, object] | None = None


class AnalysisResult(BaseModel):
    status: AnalysisStatus
    session_id: str
    answer: str
    confidence: float = Field(ge=0, le=1)
    metrics: list[Metric] = Field(default_factory=list)
    evidence: list[EvidenceRegion] = Field(default_factory=list)
    trace: list[TraceStep] = Field(default_factory=list)
    mode: DataMode = DataMode.DEVELOPMENT


class ChangeDetectionInput(BaseModel):
    aoi: AOI
    earlier_date: date
    later_date: date
    imagery: ImageryResult


class ChangeDetectionOutput(BaseModel):
    regions: list[EvidenceRegion]
    raw_detection_count: int
    detector: str
    mode: DataMode
    detector_metadata: dict[str, Any] = Field(
        default_factory=dict,
        description="Detector-level metadata (method, threshold, scene ids, etc.)",
    )


class FetchImageryInput(BaseModel):
    request: ImageryRequest


class FetchImageryOutput(BaseModel):
    result: ImageryResult


class GenerateEvidenceInput(BaseModel):
    query: str
    imagery: ImageryResult
    fused_regions: list[EvidenceRegion]
    fusion_metadata: dict[str, Any] = Field(default_factory=dict)


class FuseEvidenceInput(BaseModel):
    cva_detections: ChangeDetectionOutput
    semantic: SemanticAnalysisOutput | None = None
    sar_detections: ChangeDetectionOutput | None = None


class FuseEvidenceOutput(BaseModel):
    regions: list[EvidenceRegion]
    fusion_metadata: dict[str, Any] = Field(default_factory=dict)


class DetectSARChangeInput(BaseModel):
    aoi: AOI
    earlier_date: date
    later_date: date
    preferences: ImageryPreferences = Field(default_factory=ImageryPreferences)


class GenerateEvidenceOutput(BaseModel):
    regions: list[EvidenceRegion]
    metrics: list[Metric]
    confidence: float


class SemanticClaim(BaseModel):
    """Structured semantic claim produced by a SemanticAnalyzer."""

    claim_type: str
    region_id: str
    confidence: float = Field(ge=0, le=1)
    metrics: list[Metric] = Field(default_factory=list)
    metadata: dict[str, Any] = Field(default_factory=dict)


class SemanticAnalysisInput(BaseModel):
    aoi: AOI
    earlier_date: date
    later_date: date
    imagery: ImageryResult
    change_regions: list[EvidenceRegion]
    analysis_profile: str


class SemanticAnalysisOutput(BaseModel):
    regions: list[EvidenceRegion]
    claims: list[SemanticClaim] = Field(default_factory=list)
    analyzer: str
    mode: DataMode
    analyzer_metadata: dict[str, Any] = Field(default_factory=dict)
